crates/recursor/src/recursor.rs

```diff
@@ -466,3 +490,118 @@ fn recursor_opts() -> ResolverOpts {
 
     options
 }
+
+/// Bailiwick/sub zone checking.
+///
+/// # Overview
+///
+/// This function checks that two host names have a parent/child relationship.
+///
+/// A resolver should not return answers outside of its delegated authority -- if we receive a delegation from the root servers for
+/// "example.com", that server should only return answers related to example.com or a sub-domain thereof.  Note that record data may point
+/// to out-of-bailwick records (e.g., example.com could return a CNAME record for www.example.com that points to example.cdnprovider.net,)
+/// but it should not return a record name that is out-of-bailiwick (e.g., we ask for www.example.com and it returns www.otherdomain.com.)
+///
+/// Out-of-bailiwick responses have been used in cache poisoning attacks.
+///
+/// ## Examples
+///
+/// | Parent       | Child                | Expected Result                                                  |
+/// |--------------|----------------------|------------------------------------------------------------------|
+/// | .            | com.                 | In-bailiwick (true)                                              |
+/// | com.         | example.net.         | Out-of-bailiwick (false)                                         |
+/// | example.com. | www.example.com.     | In-bailiwick (true)                                              |
+/// | example.com. | www.otherdomain.com. | Out-of-bailiwick (false)                                         |
+/// | example.com  | www.example.com.     | Out-of-bailiwick (false, note the parent is not fully qualified) |
+///
+/// # Implementation Notes
+///
+/// * If the caller doesn't provide a parent at all, we'll return false.
+/// * ".".split('.') will evaluate to ["", ""], which for our purposes doesn't work -- that's the reason for the ends_with tests and
+/// manually pushing the root entry onto the parent/child vectors, as well as skipping the first list element if the name was already
+/// fully qualified.
+///
+/// # References
+///
+/// * [RFC 8499](https://datatracker.ietf.org/doc/html/rfc8499) -- DNS Terminology (see page 25)
+/// * [The Hitchiker's Guide to DNS Cache Poisoning](https://www.cs.utexas.edu/%7Eshmat/shmat_securecomm10.pdf) -- for a more in-depth
+/// discussion of DNS cache poisoning attacks, see section 4, specifically, for a discussion of the Bailiwick rule.
+fn is_subzone(parent: Name, child: Name) -> bool {
```

Comment:
Sorry, late comment here, did you look at `Name::zone_of` https://docs.rs/hickory-proto/latest/hickory_proto/rr/domain/struct.Name.html#method.zone_of, and was that not applicable to this use case? I just wanted to double check. If it's not complete, should we fix that method and use it instead?

Reply:
I didn't see that function before, but in testing it fails on my mixed fully/partially qualified domain tests (i.e., my function does not consider "example.com." to be a child of "com", but the Name::zone_of function does.)

Reply:
So it looks like changing Name::zone_of will cause other problems, in particular, the zone matching logic in the resolver, where the resolver may receive a request to resolve a partially qualified name and need to match that against a zone list that is fully qualified.  What I did was change my is_subzone function to have two tests to handle the special cases for bailiwick checking and then call Name::zone_of.  This simplifies this patch and avoids most code duplication (and string conversion) by using the existing function.

I thought about dropping the is_subzone function entirely and doing my extra tests inline, but I do like having tests for this, and I think the function makes those tests easier to run and be confident the subzone logic is being tested and not some other aspect of, say, the lookup function.

Let me know your thoughts.

Reply:
This looks really good. I see your points, but think this is definitely better as it reduces duplication. Thanks for making that change.